classic/create.go

```diff
@@ -0,0 +1,254 @@
+ *
+ */
+
+package classic
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/ubuntu-core/snappy/arch"
+	"github.com/ubuntu-core/snappy/dirs"
+
+	"github.com/ubuntu-core/snappy/helpers"
+	"github.com/ubuntu-core/snappy/progress"
+	"github.com/ubuntu-core/snappy/release"
+)
+
+var (
+	lxdBaseURL   = "https://images.linuxcontainers.org"
+	lxdIndexPath = "/meta/1.0/index-system"
+)
+
+func findDownloadPathFromLxdIndex(r io.Reader) (string, error) {
+	arch := arch.UbuntuArchitecture()
+	lsb, err := release.ReadLsb()
+	if err != nil {
+		return "", err
+	}
+	release := lsb.Codename
+
+	needle := fmt.Sprintf("ubuntu;%s;%s;default", release, arch)
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		if strings.HasPrefix(scanner.Text(), needle) {
+			l := strings.Split(scanner.Text(), ";")
+			if len(l) < 6 {
+				return "", fmt.Errorf("can not find download path in %s", scanner.Text())
+			}
+			return l[5], nil
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return "", err
+	}
+
+	return "", fmt.Errorf("needle %s not found", needle)
+}
+
+func findDownloadURL() (string, error) {
+	resp, err := http.Get(lxdBaseURL + lxdIndexPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to downlaod lxdIndexUrl: %s", err)
+	}
+	defer resp.Body.Close()
+
+	dlPath, err := findDownloadPathFromLxdIndex(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	url := fmt.Sprintf("%s%s%s", lxdBaseURL, dlPath, "rootfs.tar.xz")
+
+	return url, nil
+}
+
+func downloadFile(url string, pbar progress.Meter) (string, error) {
+	name := "classic"
+
+	w, err := ioutil.TempFile("", name)
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		if err != nil {
+			os.Remove(w.Name())
+		}
+	}()
+	defer w.Close()
+
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return "", err
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return "", fmt.Errorf("failed to download %s: %v", url, resp.StatusCode)
+	}
+
+	if pbar != nil {
+		pbar.Start(name, float64(resp.ContentLength))
+		mw := io.MultiWriter(w, pbar)
+		_, err = io.Copy(mw, resp.Body)
+		pbar.Finished()
+	} else {
+		_, err = io.Copy(w, resp.Body)
+	}
+
+	return w.Name(), err
+}
+
+var policyRc = []byte(`
+#!/bin/sh
+while true; do
+    case "\$1" in
+      -*) shift ;;
+      makedev) exit 0;;
+      x11-common) exit 0;;
+      *) exit 101;;
+    esac
+done
+`)
+
+var runInChroot = func(chroot string, cmd ...string) error {
+	cmdArgs := []string{"chroot", chroot}
+	cmdArgs = append(cmdArgs, cmd...)
+
+	if output, err := exec.Command(cmdArgs[0], cmdArgs[1:]...).CombinedOutput(); err != nil {
+		return fmt.Errorf("failed to run %s in chroot %s: %s (%s)", cmd, chroot, err, output)
+	}
+
+	return nil
+}
+
+var newProgress = func() progress.Meter {
```

Comment:
Silly me, indeed :) However looking at this again I noticed that I should pass the progress.Meter into Create() instead so that e.g. webdm can pass its own progress. I did this tiny refactor now.